msgpack/shared/src/main/scala/fs2/data/msgpack/high/SerializerInstances.scala

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2024 fs2-data Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package data
+package msgpack
+package high
+
+import scodec.bits.*
+
+import low.MsgpackItem
+import scala.annotation.tailrec
+
+private[high] trait SerializerInstances extends internal.PlatformSerializerInstances {
+  def right1(x: MsgpackItem) = Right(Chunk(x))
+
+  private def countBytesInt(x: Int): Int = {
+    @tailrec
+    def go(current: Int, bytes: Int): Int =
+      if (current == 0)
+        bytes
+      else
+        go(current >>> 8, bytes + 1)
+
+    if (x == 0)
+      1
+    else
+      go(x, 0)
+  }
+
+  private def countBytesLong(x: Long): Int = {
+    @tailrec
+    def go(current: Long, bytes: Int): Int =
+      if (current == 0)
+        bytes
+      else
+        go(current >>> 8, bytes + 1)
+
+    if (x == 0)
+      1
+    else
+      go(x, 0)
+  }
+
+  /* Numeric types
+   *
+   * We downcast larger types because we want squeeze the number into as little bytes as possible.
+   * We could just drop leading zeros (which we do in the else branch), but that won't work
+   * for negative numbers.
+   */
+  implicit val byteSerializer: MsgpackSerializer[Byte] = x => right1(MsgpackItem.SignedInt(ByteVector.fromByte(x)))
+
+  implicit val shortSerializer: MsgpackSerializer[Short] = { x =>
+    if (x.isValidByte)
+      byteSerializer(x.toByte)
+    else {
+      val masked = x & 0xff
+      val item =
+        if (masked == x)
+          MsgpackItem.UnsignedInt(ByteVector.fromShort(x, 1))
+        else
+          MsgpackItem.SignedInt(ByteVector.fromShort(x))
+      right1(item)
+    }
+  }
+
+  implicit val intSerializer: MsgpackSerializer[Int] = { x =>
+    if (x.isValidShort)
+      shortSerializer(x.toShort)
+    else {
+      val nbytes = countBytesInt(x)
+      val bv = ByteVector.fromInt(x, nbytes)
+      val item =
+        if (nbytes == 2)
```

Comment:
I (and especially my future self) would appreciate a few comments on why the magic `2` (bytes) allows for `UnsignedInt` here. Similarly, the masking in `shortSerializer` could use a short explanation.